app/Models/SizeVariant.php

```diff
@@ -0,0 +1,222 @@
+	{
+		parent::__construct($attributes);
+		$this->registerObserver(SizeVariantObserver::class);
+	}
+
+	/**
+	 * This model has no own timestamps as it is inseparable bound to its
+	 * parent {@link \App\Models\Photo} and uses the same timestamps.
+	 *
+	 * @var bool
+	 */
+	public $timestamps = false;
+
+	/**
+	 * List of those object relations whose timestamps shall be updated when
+	 * an object of this class is modified.
+	 *
+	 * @var string[] list of object relations
+	 */
+	protected $touches = ['photo'];
+
+	protected $casts = [
+		'full_path' => MustNotSetCast::class . ':short_path',
+		'url' => MustNotSetCast::class . ':short_path',
+		'width' => 'integer',
+		'height' => 'integer',
+		'size_variant' => 'integer',
+	];
+
+	/**
+	 * @var string[] The list of attributes which exist as columns of the DB
+	 *               relation but shall not be serialized to JSON
+	 */
+	protected $hidden = [
+		'id', // irrelevant, because a size variant is always serialized as an embedded object of its photo
+		'photo', // see above and otherwise infinite loops will occur
+		'photo_id', // see above
+		'short_path',  // serialize url instead
+		'sym_links', // don't serialize relation of symlinks
+	];
+
+	/**
+	 * @var string[] The list of "virtual" attributes which do not exist as
+	 *               columns of the DB relation but which shall be appended to
+	 *               JSON from accessors
+	 */
+	protected $appends = [
+		'url',
+	];
+
+	/**
+	 * Returns the association to the photo which this size variant belongs
+	 * to.
+	 *
+	 * @return BelongsTo
+	 */
+	public function photo(): BelongsTo
+	{
+		return $this->belongsTo(Photo::class);
+	}
+
+	/**
+	 * Returns the association to the symbolics links which point to this
+	 * size variant.
+	 *
+	 * @return HasMany
+	 */
+	public function sym_links(): HasMany
+	{
+		return $this->hasMany(SymLink::class);
+	}
+
+	/**
+	 * Accessor for the "virtual" attribute {@link SizeVariant::$url}.
+	 *
+	 * This is more than a simple convenient method which wraps
+	 * {@link SizeVariant::$short_path} into
+	 * {@link \Illuminate\Support\Facades\Storage::url()}.
+	 * Based on the current application settings and the authenticated user,
+	 * this method returns a URL to a short-living symbolic link instead a
+	 * direct URL to the actual size variant, if the underlying storage
+	 * provides symbolic links.
+	 *
+	 * @return string the url of the size variant
+	 */
+	public function getUrlAttribute(): string
+	{
+		if (
+			AccessControl::is_current_user($this->photo->owner_id) ||
+			Configs::get_value('SL_enable', '0') == '0'
+		) {
+			return Storage::url($this->short_path);
+		}
+
+		// In order to allow a grace period, we create a new symbolic link,
+		// if the most recent existing link has reached 2/3 of its lifetime
+		$maxLifetime = intval(Configs::get_value('SL_life_time_days', '3')) * 24 * 60 * 60;
+		$gracePeriod = $maxLifetime / 3;
+
+		if (Storage::getDefaultDriver() == 's3') {
```

Comment:
This code has always been there.

Actually, the use of the `Storage` interface is a mess. And no, at the current state Lychee does not support non-local storage, because there are still too many places which assume that the storage is local. These methods call `->full_path` (which only makes sense for local storage) and then use some low-level PHP function such as `@move` to do something with the file.

At the moment the code is a chimera. Many code snippets claim to be agnostic to wether `Storage` is local or non-local, but then they call a function which implicitly assumes otherwise.

After this re-factoring (and the re-factoring of the albums) this would be the next but one re-factoring. The main show stopper are the Exif Tool, FFmpeg and Imagick. At the moment the Lychee own wrappers around these tools are written such that they assume absolute paths to files on local storage. However, I have already checked that these tools actually can also handle PHP resource pointers and `Storage` provides an interface for reading/writing from/to resource pointers. So there is not technical obstacle.

There is still so much to clean up. My growing list of things that needs to be cleaned up are

- refactoring of the albums class inheritance
- clean up of the `Configs` class (see some other comment here`
- clean up of the way how `Storage` is used

I don't know who was the culprit responsible for this mess. :smirk: :smirk: